vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/BigDecimalField.java

```diff
@@ -224,9 +224,15 @@ public void setValue(BigDecimal value) {
                 && Objects.equals(value, getEmptyValue())
                 && isInputValuePresent()) {
             // Clear the input element from possible bad input.
-            getElement().executeJs("this.inputElement.value = ''");
+            getElement().executeJs("this._inputElementValue = ''");
             getElement().setProperty("_hasInputValue", false);
             fireEvent(new ClientValidatedEvent(this, false));
+        } else {
+            // Restore the input element's value in case it was cleared
+            // in the above branch. That can happen when setValue(null)
+            // and setValue(...) are subsequently called within one round-trip
+            // and there was bad input.
+            getElement().executeJs("this._inputElementValue = this.value");
```

Comment:
The solution could actually be much simpler if we had a `_hasBadInput` property instead of `_hasInputValue` (see a proposal in https://github.com/vaadin/web-components/issues/5639).

Example:

```java
if (Objects.equals(oldValue, getEmptyValue())
    && Objects.equals(value, getEmptyValue())
    && isBadInputPresent()) {
    getElement().executeJs("if (this._hasBadInput) this._inputElementValue = ''");
    getElement().setProperty("_hasBadInput", false);
    fireEvent(new ClientValidatedEvent(this, false));
}
```

Reply:
BigDecimalField is different from such components as TimePicker, IntegerField, etc, in the way that it doesn't clear its value when the user enters bad input. For that reason, I had to use a different approach than in [other components](https://github.com/vaadin/flow-components/pull/4828/files#diff-82671d2386a473925418e897ecccbb7c6ef2ab89bb7fe9df89b41809958244f4R260-R268). I rather restore the input element's value than prevent `executeJs` from clearing it.